components/backup-stream/src/checkpoint_manager.rs

```diff
@@ -0,0 +1,352 @@
+        let fresh_regions = self.resolvers.collect_fresh_subs();
+        let removal = self.resolvers.collect_removal_subs();
+        let checkpoints = removal
+            .into_iter()
+            .map(|sub| (sub.meta, sub.resolver.resolved_ts()))
+            .collect::<Vec<_>>();
+        self.checkpoints = checkpoints;
+        self.fresh_regions = fresh_regions;
+    }
+
+    async fn after(&mut self, task: &str, rts: u64) -> Result<()> {
+        if !self.can_advance.take().map(|f| f()).unwrap_or(true) {
+            let cp_now = self
+                .meta_cli
+                .get_local_task_checkpoint(&task)
+                .await
+                .context(format_args!(
+                    "during checking whether we should skip advencing ts to {}.",
+                    rts
+                ))?;
+            // if we need to roll back checkpoint ts, don't prevent it.
+            if rts >= cp_now.into_inner() {
+                info!("skipping advance checkpoint."; "rts" => %rts, "old_rts" => %cp_now);
+                return Ok(());
+            }
+        }
+        self.base.after(task, rts).await?;
+        // Optionally upload the region checkpoint.
+        // Unless in some exteme condition, skipping upload the region checkpoint won't lead to data loss.
+        if let Err(err) = self
+            .meta_cli
+            .upload_region_checkpoint(&task, &self.checkpoints)
+            .await
+        {
+            err.report("failed to upload region checkpoint");
+        }
+        self.meta_cli
+            .clear_region_checkpoint(&task, &self.fresh_regions)
+            .await
+            .context(format_args!("on clearing the checkpoint for task {}", task))?;
+        Ok(())
+    }
+}
+
+pub struct CheckpointV3FlushObserver<S, O> {
+    sched: Scheduler<Task>,
+    meta_cli: MetadataClient<S>,
+    checkpoints: Vec<(Region, TimeStamp)>,
+
+    /// We should modify the rts (the local rts isn't right.)
+    /// This should be a BasicFlushObserver or something likewise.
+    baseline: O,
+}
+
+impl<S, O> CheckpointV3FlushObserver<S, O> {
+    pub fn new(sched: Scheduler<Task>, meta_cli: MetadataClient<S>, baseline: O) -> Self {
+        Self {
+            sched,
+            meta_cli,
+            checkpoints: vec![],
+            baseline,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<S, O> FlushObserver for CheckpointV3FlushObserver<S, O>
+where
+    S: MetaStore + 'static,
+    O: FlushObserver + Send,
+{
+    async fn before(&mut self, checkpoints: Vec<(Region, TimeStamp)>) {
+        self.checkpoints = checkpoints;
+    }
+
+    async fn after(&mut self, task: &str, _rts: u64) -> Result<()> {
+        let t = Task::RegionCheckpointsOp(RegionCheckpointOperation::Update(std::mem::take(
+            &mut self.checkpoints,
+        )));
+        try_send!(self.sched, t);
+        let global_checkpoint = self.meta_cli.global_checkpoint_of_task(task).await?;
```

Comment:
we already called `global_checkpoint_of_task` in `rewrite_resolved_ts`, can we use that global checkpoint to reduce one time rpc invoke?